src/supernode/src/main/java/com/dragonflyoss/dragonfly/supernode/service/repair/FileMd5NotMatchHandler.java

```diff
@@ -0,0 +1,135 @@
+package com.dragonflyoss.dragonfly.supernode.service.repair;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.RandomAccessFile;
+import java.nio.file.Path;
+import java.security.MessageDigest;
+import java.util.List;
+
+import com.dragonflyoss.dragonfly.supernode.common.domain.ClientErrorInfo;
+import com.dragonflyoss.dragonfly.supernode.common.domain.FileMetaData;
+import com.dragonflyoss.dragonfly.supernode.common.enumeration.ClientErrorType;
+import com.dragonflyoss.dragonfly.supernode.common.util.RangeParseUtil;
+import com.dragonflyoss.dragonfly.supernode.service.cdn.FileMetaDataService;
+import com.dragonflyoss.dragonfly.supernode.service.cdn.util.PathUtil;
+import com.dragonflyoss.dragonfly.supernode.service.timer.DataGcService;
+
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.codec.binary.Hex;
+import org.apache.commons.codec.digest.DigestUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+/**
+ * This class only handle the errors that are caused by downloading from supernode.
+ * It checks the piece md5 on local cache, verify it and delete the related task
+ * if the md5 is not expected.
+ *
+ * @author lowzj
+ */
+@Component
+@Slf4j
+public class FileMd5NotMatchHandler extends BaseClientErrorHandler {
+
+    @Autowired
+    private FileMetaDataService fileMetaDataService;
+
+    @Autowired
+    private DataGcService gcService;
+
+    @Override
+    ClientErrorType errorType() {
+        return ClientErrorType.FILE_MD5_NOT_MATCH;
+    }
+
+    @Override
+    void doHandle(final ClientErrorInfo info) {
+        boolean gc = false;
+        String md5FromMeta = getMd5FromMeta(info.getTaskId(), info.getRange());
+        String localMd5 = getMd5(info.getTaskId(), info.getRange());
+        if (StringUtils.isNotBlank(localMd5)) {
+            if (!StringUtils.equals(md5FromMeta, localMd5) // data in supernode is not right
+                && !StringUtils.equals(localMd5, info.getExpectedMd5())) {
```

Comment:
i think if data in supernode is not right then we will remove the task from the super.
is condition !StringUtils.equals(localMd5, info.getExpectedMd5()) necessary?

Reply:
Before supernode completing downloading file from remote server, the meta data is not written into local disk, so the `md5FromMeta` may be null.